src/Surfnet/AzureMfa/Application/Service/AzureMfaService.php

```diff
@@ -0,0 +1,177 @@
+<?php declare(strict_types=1);
+
+/**
+ * Copyright 2019 SURFnet B.V.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Surfnet\AzureMfa\Application\Service;
+
+use Exception;
+use RobRichards\XMLSecLibs\XMLSecurityKey;
+use SAML2\Assertion;
+use SAML2\Certificate\PrivateKeyLoader;
+use SAML2\Configuration\PrivateKey;
+use Surfnet\SamlBundle\Entity\IdentityProvider;
+use Surfnet\SamlBundle\Entity\ServiceProvider;
+use Surfnet\SamlBundle\Http\PostBinding;
+use Surfnet\SamlBundle\SAML2\AuthnRequest;
+use Surfnet\SamlBundle\SAML2\AuthnRequestFactory;
+use Symfony\Component\HttpFoundation\RedirectResponse;
+use Symfony\Component\HttpFoundation\Request;
+
+class AzureMfaService
+{
+
+    /**
+     * @var string
+     */
+    private $publicKey;
+    /**
+     * @var string
+     */
+    private $privateKey;
+    /**
+     * @var PostBinding
+     */
+    private $postBinding;
+
+    public function __construct(
+        PostBinding $postBinding
+    ) {
+        $this->postBinding = $postBinding;
+
+        // Todo: make keys configurable
+        $this->publicKey = __DIR__ . '/../../../../../vendor/surfnet/stepup-saml-bundle/src/Resources/keys/development_publickey.cer';
+        $this->privateKey = __DIR__ . '/../../../../../vendor/surfnet/stepup-saml-bundle/src/Resources/keys/development_privatekey.pem';
+    }
+
+    /**
+     * @param string $nameId
+     * @return RedirectResponse
+     * @throws Exception
+     */
+    public function handleAuthnRequestToAdfs(string $nameId): string
+    {
+        $authnRequest = AuthnRequestFactory::createNewRequest($this->getServiceProvider(), $this->getIdentityProvider());
```

Comment:
My primary goal was to implement the service without having to bother about configuration and state while still be able to utilize the mock MFA client in our development environment. The IdP through a factory seems a very good idea because we would be able to mock this really easily in our development boxes so we could utilize the mock client.